File: 3.x.x/crates-utopia/emails/src/canonicals/protonmail.rs
use crate::canonicals::provider::{remove_plus_addressing, to_lower_case, Canonical, Provider};
use crate::EmailError;

/// Proton Mail canonicalization (PHP `Protonmail`).
///
/// Keeps the original domain when it is supported; otherwise `protonmail.com`.
#[derive(Debug, Clone, Copy, Default)]
pub struct Protonmail;

const SUPPORTED_DOMAINS: &[&str] = &["protonmail.com", "proton.me", "pm.me", "protonmail.ch"];
const CANONICAL_DOMAIN: &str = "protonmail.com";

impl Provider for Protonmail {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, domain: &str) -> Result<Canonical, EmailError> {
        let normalized = to_lower_case(local);
        let normalized = remove_plus_addressing(&normalized);
        let canonical_domain = if SUPPORTED_DOMAINS.contains(&domain) {
            domain
        } else {
            CANONICAL_DOMAIN
        };
        Ok(Canonical {
            local: normalized,
            domain: canonical_domain.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Emails/Canonicals/Providers/Protonmail.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * ProtonMail
 *
 * Handles ProtonMail email normalization
 * - Removes plus addressing (subaddress) from local part
 * - Preserves dots in local part
 * - Does not normalize domains
 *
 * Docs: https://proton.me/support/creating-aliases#+Aliases
 */
class Protonmail extends Provider
{
    private const SUPPORTED_DOMAINS = ['protonmail.com', 'proton.me', 'pm.me', 'protonmail.ch'];

    private const CANONICAL_DOMAIN = 'protonmail.com';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Remove plus addressing (subaddress) - everything after +
        $normalizedLocal = $this->removePlusAddressing($normalizedLocal);

        // protonmail.ch, protonmail.com - not subaddress, just different options during sign up
        // pm.me - technically subaddress, but costs monthly fee, and gives just +1 email. Costly already, no need to block. People get it for shorter email to type it quicker anyway
        return [
            'local' => $normalizedLocal,
            'domain' => \in_array($domain, self::SUPPORTED_DOMAINS, true) ? $domain : self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
